Fix color of image masks inside uncolored patterns

janpe2 opened the pull request:
Fixes #8741 
Fixes #3226
Fixes #8860

This is a patch for uncolored (`/PaintType 2`) tiling patterns that contain image masks. When uncolored patterns are used as a fill or stroke, the operator `scn` or `SCN` specifies the color of the pattern. If the pattern contains an image mask, PDF.js incorrectly draws it using black instead of the color specified by `scn` or `SCN`. 

In my patch, I assign the proper color to the `fillColor` property of the `CanvasExtraState` object. That property is used as the color of image masks in functions `paintImageMaskXObject...()` in canvas.js.

yurydelendik commented on src/display/pattern_helper.js:
```diff
@@ -360,7 +360,7 @@ var TilingPattern = (function TilingPatternClosure() {
       graphics.groupLevel = owner.groupLevel;
 
-      this.setFillAndStrokeStyleToContext(tmpCtx, paintType, color);
+      this.setFillAndStrokeStyleToContext(tmpCtx, paintType, color, graphics);
```
let's change signature to setFillAndStrokeStyleToGraphics(graphic, paintType, color) -- `context` is accessible via graphics.ctx.

yurydelendik commented on src/display/pattern_helper.js:
```diff
@@ -412,6 +413,8 @@ var TilingPattern = (function TilingPatternClosure() {
             context.fillStyle = cssColor;
             context.strokeStyle = cssColor;
+            // Set color needed by image masks (fixes issues 3226 and 8741).
+            graphics.current.fillColor = cssColor;
```
do we need to do that for `this.current.strokeColor =`? do we need to do the same for PaintType.COLORED case?

janpe2 commented:
> do we need to do that for this.current.strokeColor =?

I don't know if any PDF would need this but I added it, just to be sure.

> do we need to do the same for PaintType.COLORED case?

I added lines similar to the other case. Colored tiling patterns are a bit different: now the `color` argument in `setFillAndStrokeStyleToContext()` is meaningless (or null) because the PDF operators scn and SCN have no color operands.

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
/botio test

Snuffleupagus commented:
> Fixes #8741, #3226.

Would it be possible to include the PDF file from either of those issues as a *linked* test-case, to prevent future regressions?

janpe2 commented:
I added the test case.

By the way, my patch seems to fix also issue #8860.

Snuffleupagus commented:
> I added the test case.

Thank you; one question though, wouldn't it be enough to just test the first page though?
If so, please add `"lastPage": 1,` to entry in `test_manifest.json`

Snuffleupagus commented:
/botio test

Snuffleupagus commented:
@janpe2 Looking at the test logs above, it appears that the server prevented the PDF file from being downloaded:
```
Downloading http://arxiv.org/pdf/0705.4298.pdf to pdfs/issue3226.pdf...
Error during downloading of http://arxiv.org/pdf/0705.4298.pdf: HTTP 403
WARNING: File was not downloaded. See "pdfs/issue3226.pdf.error" file.
Unable to verify the checksum for the files that are used for testing.
Please re-download the files, or adjust the MD5 checksum in the manifest for the files listed above.
```

Would it be possible to link to the PDF file through e.g. http://archive.org/web/ instead, or do we need to use another PDF file for the test-case?

timvandermeij commented:
/botio test

janpe2 commented:
What are those failed regression tests? Should I be worried about them?

timvandermeij commented:
The problem is that the old file is not replaced. @yurydelendik Could you remove the old `issue3226.pdf` file from both bots so we can try again?

Snuffleupagus commented:
> The problem is that the old file is not replaced. @yurydelendik Could you remove the old (cached) issue3226.pdf file from both bots so we can try again?

I don't think that's really the issue at play here, since the (Internet Archive) link does *not* actually point to a PDF file.
If you open https://web.archive.org/web/20170830175223/http://arxiv.org/pdf/0705.4298.pdf directly in a browser, you'll get the same thing that the test-suite gets when downloading the file: a HTML page which references a PDF file.

It seems to me that linking to this particular PDF file probably will not work, and that the easiest solution here would be to try and use a PDF file from one of the other referenced issues as a test-case instead.

janpe2 committed:
Fix color of image masks inside uncolored patterns

Snuffleupagus commented:
/botio test

Snuffleupagus reviewed (approved):
Looks good, thank you for the patch!

Snuffleupagus commented:
/botio makeref
